### PR TITLE
MM-21336: Improve channel rename restrictions

### PR DESCRIPTION
**Pull request by agnivade**
#### Summary

Non-direct channels should not be of the form "userID__userID".

#### Ticket Link

https://mattermost.atlassian.net/browse/MM-21336

### CODE REVIEW

**Comment by hanzei** [comment]
Just to double check: This is a breaking change for both the REST API and the plugin API, right?

**Comment by jespino** [comment]
The code looks good to me, but I think we are being too restrictive here, for example, `x__x` is going to be an invalid channel name. I would suggest to check the length of the channel name too, so, something like

```go
userIds := strings.Split(channelName)
if len(userIds) == 2 && model.IsValidId(userIds[0]) && model.IsValidId(userIds[1]) {
    ...
}
```

**Comment by agnivade** [comment]
Actually, this is the same check in `CreateChannelWithUser`. Hence I just replicated that. Do you want me to change that too ?

**Comment by jespino** [comment]
Yes, I think that is what makes sense, there is no reason for anybody to change the name of a direct channel to something different to userId__userId, so this rules must be safe.

**Comment by agnivade** [comment]
Actually I refrained from changing `CreateChannelWithUser`. Because the error message has translations throughout all languages. And that would need to be changed too. So just went with this change.

Should this have a PM review too to check the error message ?

File: app/channel.go [hunk 1]
```diff
@@ -508,7 +508,16 @@ func (a *App) GetGroupChannel(userIds []string) (*model.Channel, *model.AppError
...
 }
 
+// UpdateChannel updates a given channel by its Id. It also publishes the CHANNEL_UPDATED event.
 func (a *App) UpdateChannel(channel *model.Channel) (*model.Channel, *model.AppError) {
+	userIds := strings.Split(channel.Name, "__")
+	if channel.Type != model.CHANNEL_DIRECT &&
+		len(userIds) == 2 &&
+		model.IsValidId(userIds[0]) &&
+		model.IsValidId(userIds[1]) {
+		return nil, model.NewAppError("UpdateChannel", "api.channel.update_channel.invalid_character.app_error", nil, "", http.StatusBadRequest)
```
Comment (mkraft) [inline app/channel.go]:
```suggestion
		return nil, model.NewAppError("App.UpdateChannel", "api.channel.update_channel.invalid_character.app_error", nil, "", http.StatusBadRequest)
```

**Comment by agnivade** [comment]
> ...but that's not what the code does. Which one is incorrect?

Ooh yeah, it got changed in the review. Thanks.

File: app/channel.go (hunk 1 above)
Comment (agnivade) [reply app/channel.go]: The prefix "App." is nowhere used in this file. I stuck with consistency. Is it fine to prefix "App." ? Then should we be doing it everywhere else too ?

Comment (mkraft) [reply app/channel.go]: Ah OK, I guess it's not being done for the App methods for some reason. Disregard my suggestion.

**Comment by agnivade** [comment]
Jesus is on leave. Requesting Eli to take a look.

File: i18n/en.json [hunk 2]
```diff
@@ -371,6 +371,10 @@
     "id": "api.channel.update_channel.deleted.app_error",
     "translation": "The channel has been archived or deleted"
   },
+  {
+    "id": "api.channel.update_channel.invalid_character.app_error",
+    "translation": "Invalid channel name for non-direct channels. Cannot have user id."
```
Comment (wiersgallak) [inline i18n/en.json]: Recommend updating to be more clear and consistent in identifying the channel as non-direct **message** channel:  

"Invalid channel name.  User ids are not permitted in channel name for non-direct message channels."

**Comment by DHaussermann** [comment]
@agnivade thanks for the fix on this. The issue as described is resolved. 

I did find a bug that is a variation of this issue. On create we validate against `<userid1>__<userid2>` but, I can create a channel named `<userid1> <userid2>` the app then creates a channel with a URL of `<userid1>--<userid2>`. This channel then exhibits some of the same strange behavior. It cannot be selected from the LHS. 
Please advise if this requires a separate issue or if it can also be resolved in this PR.

**Comment by agnivade** [comment]
@DHaussermann - I was not able to reproduce such behavior. I created a channel with "userid1 userid2". The app created url as "userid1-userid2". Note the single dash, not double dash. I don't see any issues in the app.

channels

**Comment by agnivade** [comment]
I do see that deliberately renaming to "userid1--userid2" manifests the issue.

**Comment by agnivade** [comment]
Okay so this is a different issue, something related to preference categories storing the category with "--userid" being appended at the end. The webapp thinks it is a user and makes a request to "/users/userid" to get the user details. And upon not finding it, redirects to town square.

```
DEBU[2020-01-03T11:26:55.5854347+05:30] not found handler triggered                   caller="web/web.go:66" code=404 ip="::1" path=/api/v4/users/6zii9a9g6pruzj451x3esok54h--wr4j4g8zqtnhmkw771pfpynqwo
```

I would suggest to file a separate issue for this to investigate further.

**Comment by agnivade** [comment]
@amyblais - Can you address @hanzei's concerns that it may be a breaking change ? Should there be release notes for this ?

**Comment by amyblais** [comment]
@agnivade Yes it should be documented if it's a breaking change. I don't know if Platform team or Hanzei would confirm if it is.

**Comment by agnivade** [comment]
The question is how would you classify what a breaking change is. This was a bug which allowed to create bad channels. This change prevents that. If this is considered a breaking change, then yes it should be documented.

/cc @grundleborg

**Comment by grundleborg** [comment]
I would not consider this a breaking change. The behaviour it fixes is clearly a bug - nobody could be relying on that behaviour as it results in the channels getting broken. It is a behaviour that has always been clearly unsupported, because it breaks everything.

**Pull request merged**